### DISCUSSION TITLE
Bring features from object storage to network storage

### DISCUSSION THREAD

**Discussion by alexq555** [ClickHouse/ClickHouse#33998] [Ideas]
AWS EFS can be a good data storage solution for ClickHouse, see https://github.com/ClickHouse/ClickHouse/issues/31113 , namely high throughput speeds for stores in the 10s-100s of TB range, automatic sizing and lifecycle management. Between local and object storage network shares have not been a huge focus in CH up until this point.

It seems a bit counterintuitive that despite storage being shared, often with very high redundancy (11 9’s for AWS EFS) replication is still required between multiple servers. Users who have provisioned enough throughput may not need the costly redundancy that replication brings.

Two features developed for S3/object stores therefore could be perfect for network shares as well, namely zero copy replication and disk cache. 

A new disk type nfs could be added that largely extends the object storage types such as s3. If shares are mounted on the same mount point then zero replication could be enabled by storing the pointers to the nfs files locally, just like with object stores. A nice to have feature would be to keep the directory structure matching the partition and file names (even if we need to add some random extra characters to avoid conflicts on shared storage), which would make life easier if ever we wanted to get something back from backup, 

If users wanted to use network storage individually and replicate they could continue to declare them as simple disks, in other words only actively identifying the storage as nfs would it turn it into something more like object-storage. 

In the end, with perhaps just a few tweaks nfs could fit into the remote_filesystem framework that has been expanding beyond just S3 to other kinds of stores recently. 

This could be another option for users wishing to decouple compute from storage and flexibly provision their compute requirements independent of storage.

**Comment by alexey-milovidov** [comment]
We are experimenting with running almost ordinary MergeTree on top of the same data directory in shared filesystem.
It requires very minimal amount of coordination compared to ReplicatedMergeTree.
Looks very promising.

**Comment by alexq555** [comment]
That sounds exciting indeed!

On the caching point specifically, I looked through the Local Cache for remote filesystem PRs https://github.com/ClickHouse/ClickHouse/pull/33717 (@kssenii / @alesapin ) and didn't get the impression that they're supporting anything other than remote object stores yet. 

Making that a totally generic option for all data stores, i.e. block and network storage as well, would be a chance to make full use of things like AWS local SSD disks, which are not safe enough to use as data stores otherwise, given that they disappear when instances are terminated. It may seem redundant but EBS/remote network shares may be more throughput/IOPS limited than directly attached NVMe drives and adding that caching layer regardless of whether it's remote could make a lot of sense.

**Comment by c3mb0** [comment]
I believe this feature is greatly beneficial and needed for catching up with the current data processing trend of separating compute and storage. Google's BigQuery and Snowflake already does this, and AWS recently announced Redshift Serverless which basically decouples compute.

Being an AWS user, here are the storage options I'm really looking forward to experiment with:
- Both `io1` and `io2` EBS storages have a multi-attach option, meaning I can attach to same volume to many EC2 instances as long as they are in the same AZ (basically POSIX storage)
- FSx supports both `zfs` and `Lustre` which are multi-attachable across AZs (zfs is attached via NFS, Lustre has its own protocol)
- EFS as you mentioned is also multi-attachable across AZs (NFS)

With multi-attachable volumes, storage is essentially decoupled from compute, which would enable users to throw in as much compute as they wish. It would be great to be able to specify `allow_remote_fs_zero_copy_replication` for any/all these types of storages.

**Reply by den-crane** [reply]
it's not real decoupling, you can grow number of replicas but not number of shads.

**Reply by c3mb0** [reply]
Assuming the underlying storage provides enough resources (size, IOPS, inodes, bandwidth), is there a reason to shard the data at all?

**Comment by marising** [comment]
I am very glad that there are a lot of guys with this idea, we need ClickHouse support for NFS (Network file system), the simplest usage:

1. NFS mount to the local server and use it like local disk
2. Multiple replicas of the same shard share an NFS path and do not require data synchronization between replicas #16240

I think it's worth it if we only lose a portion of the read or write performance, and we can bring more capacity, throughput, and elasticity.

Don't know what someone is doing? When will it be available?

**Comment by maskshell** [comment]
If we consider a cloud service such as AWS, then cloud-native shared block storage is more worth considering than a network file system.

Think about Oracle ASM / Veritas CFS / GPFS / OCFS2 ...

I know Alibaba Cloud has two products, cloud-native DBFS (Database File System), and cloud-native Shared Block Storage. The first one can be used directly for the underlying database storage, and the second can be used to create shared block storage such as OCFS2.
